python/ql/src/experimental/semmle/python/frameworks/Invoke.qll

```diff
@@ -0,0 +1,150 @@
+  // invoke
+  // ---------------------------------------------------------------------------
+  /** Gets a reference to the `invoke` module. */
+  private DataFlow::Node invoke(DataFlow::TypeTracker t) {
+    t.start() and
+    result = DataFlow::importModule("invoke")
+    or
+    exists(DataFlow::TypeTracker t2 | result = invoke(t2).track(t2, t))
+  }
+
+  /** Gets a reference to the `invoke` module. */
+  DataFlow::Node invoke() { result = invoke(DataFlow::TypeTracker::end()) }
+
+  /**
+   * Gets a reference to the attribute `attr_name` of the `invoke` module.
+   * WARNING: Only holds for a few predefined attributes.
+   */
+  private DataFlow::Node invoke_attr(DataFlow::TypeTracker t, string attr_name) {
+    attr_name in ["run", "sudo", "context", "Context", "task"] and
+    (
+      t.start() and
+      result = DataFlow::importMember("invoke", attr_name)
+      or
+      t.startInAttr(attr_name) and
+      result = DataFlow::importModule("invoke")
+    )
+    or
+    // Due to bad performance when using normal setup with `invoke_attr(t2, attr_name).track(t2, t)`
+    // we have inlined that code and forced a join
+    exists(DataFlow::TypeTracker t2 |
+      exists(DataFlow::StepSummary summary |
+        invoke_attr_first_join(t2, attr_name, result, summary) and
+        t = t2.append(summary)
+      )
+    )
+  }
+
+  pragma[nomagic]
+  private predicate invoke_attr_first_join(
+    DataFlow::TypeTracker t2, string attr_name, DataFlow::Node res, DataFlow::StepSummary summary
+  ) {
+    DataFlow::StepSummary::step(invoke_attr(t2, attr_name), res, summary)
+  }
+
+  /**
+   * Gets a reference to the attribute `attr_name` of the `invoke` module.
+   * WARNING: Only holds for a few predefined attributes.
+   */
+  private DataFlow::Node invoke_attr(string attr_name) {
+    result = invoke_attr(DataFlow::TypeTracker::end(), attr_name)
+  }
+
+  /** Provides models for the `invoke` module. */
+  module invoke {
+    /** Gets a reference to the `invoke.context` module. */
+    DataFlow::Node context() { result = invoke_attr("context") }
+
+    /** Provides models for the `invoke.context` module */
+    module context {
+      /** Provides models for the `invoke.context.Context` class */
+      module Context {
+        /** Gets a reference to the `invoke.context.Context` class. */
+        private DataFlow::Node class_(DataFlow::TypeTracker t) {
+          t.start() and
+          result = DataFlow::importMember("invoke.context", "Context")
+          or
+          t.startInAttr("Context") and
+          result = invoke::context()
+          or
+          t.start() and
+          result = invoke_attr("Context")
+          or
+          exists(DataFlow::TypeTracker t2 | result = class_(t2).track(t2, t))
+        }
+
+        /** Gets a reference to the `invoke.context.Context` class. */
+        DataFlow::Node class_() { result = class_(DataFlow::TypeTracker::end()) }
```

Comment:
Argh, I'm really conflicted about the use of `class_`. I think it's an eyesore, but I can see the appeal in having a uniform way of saying "this is a class I'm modelling". Perhaps it would be better to instead use a more neutral term like `tracker`? I'm thinking especially of cases where functions are used as a proxy for classes and vice versa. Would we in these cases want to match what the underlying implementation does, or the way the API is used? If instead we just said "this is `foo::bar::Baz::trackedNode()`" or something along those lines, I think it would be more clear what's going on (and that we don't really care about what the underlying thing is, just that it's a node that refers to `foo.bar.Baz`).